src/overlays/actors/ovl_En_Clear_Tag/z_en_clear_tag.c

```diff
@@ -61,25 +76,999 @@ static ColliderCylinderInit D_809D5C40 = {
+    }
+}
+
+/**
+ * EnClear_Tag update function.
+ *
+ * ClearTag has three variants:
+ * CLEAR_TAG_CUTSCENE_ARWING which is the regular Arwing and plays a cutscene on spawn.
+ * CLEAR_TAG_ARWING which is a regular Arwing.
+ * CLEAR_TAG_LASER which are the lasers fired by the Arwing.
+ *
+ * This function controls the Arwing flying. A target position is set and the Arwing flies toward it based on
+ * calculated directions from the current position.
+ * This function fires Arwing lasers when the state is CLEAR_TAG_STATE_TARGET_LOCKED.
+ * This function controls the cutscene that plays when the Arwing has params for
+ * cutscene. The cutscene stops playing when the Arwing is a specified distance from the starting point.
+ *
+ * work[0] for Arwing in locked on mode is a timeout when the Arwing reverts back to flying mode.
+ * work[0] for Arwing in fly mode is a timeout for when you can enter locked on mode.
+ * work[0] for lasers is a cutscene timer.
+ * work[1] for Arwing is a timeout for when it will enter locked on mode.
+ * work[2] for Arwing is a cutscene timer.
+ */
+void EnClearTag_Update(Actor* thisx, GlobalContext* globalCtx) {
+    u8 hasAtHit = 0;
+    s16 i;
+    s16 xRotationTarget;
+    s16 rotationScale;
+    EnClearTag* this = THIS;
+    GlobalContext* globalCtx2 = globalCtx;
+    Player* player = PLAYER;
+    f32 vectorToTargetX;
+    f32 vectorToTargetY;
+    f32 vectorToTargetZ;
+    s16 worldRotationTargetZ;
+    s16 cutsceneTimer;
+    f32 loseTargetLockDistance;
+    s32 pad;
+    s16 worldRotationTargetY;
+    s16 worldRotationTargetX;
+    f32 targetCircleX;
+    f32 targetCircleZ;
+    f32 cutsceneCameraCircleX;
+    f32 cutsceneCameraCircleZ;
+    s16 isArwingAlive;
+    Vec3f cutsceneCameraAtTarget;
+    Vec3f cutsceneCameraEyeTarget;
+    Vec3f crashEffectLocation;
+    Vec3f crashEffectVelocity;
+    Vec3f debrisEffectAcceleration;
+
+    this->timer++;
+
+    if (this->drawMode != CLEAR_TAG_DRAWMODE_EFFECT) {
+        for (i = 0; i < 3; i++) {
+            if (this->work[i] != 0) {
+                this->work[i]--;
+            }
+        }
+
+        if (this->cutsceneTimer != 0) {
+            this->cutsceneTimer--;
+        }
+
+        switch (this->state) {
+            case CLEAR_TAG_STATE_DEMO:
+            case CLEAR_TAG_STATE_TARGET_LOCKED:
+            case CLEAR_TAG_STATE_FLYING:
+                // Check if the Arwing should crash.
+                if (this->collider.base.acFlags & 2 &&
+                    (this->collider.base.acFlags &= ~2, this->crashingTimer = 20,
+                     Actor_SetColorFilter(&this->actor, 0x4000, 255, 0, 5), this->acceleration.x = Rand_CenteredFloat(15.0f),
+                     this->acceleration.y = Rand_CenteredFloat(15.0f), this->acceleration.z = Rand_CenteredFloat(15.0f),
+                     Audio_PlayActorSound2(&this->actor, NA_SE_EN_FANTOM_THUNDER_GND), this->actor.colChkInfo.health--,
+                     (s8)this->actor.colChkInfo.health <= 0)) {
```

Comment:
OK this can't possibly be correct.

Reply:
Give me an alternative. ;)

Reply:
Using 1 or 2 goto here should work, and it would probably lead to better code (?), but yeah I can't think of any alternative that doesn't involve goto(s)

Reply:
It doesn't logically need a goto. It's logically this, but it's not OK.
```c
                if (this->collider.base.acFlags & 2) {
                     this->collider.base.acFlags &= ~2;
                     this->crashingTimer = 20;
                     Actor_SetColorFilter(&this->actor, 0x4000, 255, 0, 5);
                     this->acceleration.x = Rand_CenteredFloat(15.0f);
                     this->acceleration.y = Rand_CenteredFloat(15.0f);
                     this->acceleration.z = Rand_CenteredFloat(15.0f);
                     Audio_PlayActorSound2(&this->actor, NA_SE_EN_FANTOM_THUNDER_GND);
                     this->actor.colChkInfo.health--;
                     if ((s8)this->actor.colChkInfo.health <= 0)) {
                         // Do the rest of the things.
                     }
                }
```